pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +239,298 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+
+	// We should take the max of "seen till" offset. If the partition was lagging
+	// due to some record not being processed because of a future sample, we might be
+	// coming back to the same consume cycle again.
+	oldestSeenOffset := max(lastRec.Offset, state.OldestSeenOffset)
+	// Take the max of block max times because of the same reasons above.
+	lastBlockEnd := blockEnd
+	if lastBlockEnd.Before(state.LastBlockEnd) {
+		lastBlockEnd = state.LastBlockEnd
+	}
+
+	commit := kadm.Offset{
+		Topic:       commitRec.Topic,
+		Partition:   commitRec.Partition,
+		At:          commitRecOffset,
+		LeaderEpoch: commitRec.LeaderEpoch,
+		Metadata:    marshallCommitMeta(commitRec.Timestamp.UnixMilli(), oldestSeenOffset, lastBlockEnd.UnixMilli()),
+	}
+	if err := b.commitOffset(ctx, b.cfg.Kafka.ConsumerGroup, commit); err != nil {
+		return err
+	}
+
+	// The new lag is the distance between fully processed offsets. If the partition's cycle was chopped into sections due to lagging,
+	// the cycle will continue consuming the next portion of the lag.
+	state.Lag -= commitRecOffset - firstRec.Offset
+
+	state.OldestSeenOffset = oldestSeenOffset
+	state.LastBlockEnd = lastBlockEnd
+	state.Commit = commit
+
+	return nil
+}
+
+func (b *BlockBuilder) seekPartition(partition int32, offset int64) {
+	offsets := map[string]map[int32]kgo.EpochOffset{
+		b.cfg.Kafka.Topic: {
+			partition: {
+				Epoch:  -1, // Epoch doesn't play any role outside the group consumption.
+				Offset: offset,
+			},
+		},
+	}
+	b.kafkaClient.SetOffsets(offsets)
+}
+
+func (b *BlockBuilder) commitOffset(ctx context.Context, group string, offset kadm.Offset) error {
+	offsets := make(kadm.Offsets)
+	offsets.Add(offset)
+	if err := kadm.NewClient(b.kafkaClient).CommitAllOffsets(ctx, group, offsets); err != nil {
+		return fmt.Errorf("commit with partition %d, offset %d: %w", offset.Partition, offset.At, err)
+	}
+
+	level.Debug(b.logger).Log("msg", "successfully committed to Kafka", "partition", offset.Partition, "offset", offset.At)
+
+	return nil
+}
+
+func (b *BlockBuilder) uploadBlocks(ctx context.Context, tenantID, dbDir string, blockIDs []string) error {
+	buc := bucket.NewUserBucketClient(tenantID, b.bucket, b.limits)
+	for _, bid := range blockIDs {
+		blockDir := path.Join(dbDir, bid)
+		meta, err := block.ReadMetaFromDir(blockDir)
+		if err != nil {
+			return err
+		}
+
+		if meta.Stats.NumSamples == 0 {
+			// No need to upload empty block.
+			return nil
+		}
+
+		meta.Thanos.Source = block.BlockBuilderSource
+		meta.Thanos.SegmentFiles = block.GetSegmentFiles(blockDir)
+
+		if meta.Compaction.FromOutOfOrder() && b.limits.OutOfOrderBlocksExternalLabelEnabled(tenantID) {
+			// At this point the OOO data was already ingested and compacted, so there's no point in checking for the OOO feature flag
+			meta.Thanos.Labels[mimir_tsdb.OutOfOrderExternalLabel] = mimir_tsdb.OutOfOrderExternalLabelValue
+		}
```

Comment:
Normally we don't expect empty blocks, right?. I'd log this at info level with the ulid of the block so we can trace why there were no samples in this block (maybe all the samples were invalid? unlikely, but possible).

Reply:
I cannot figure out which piece of code you refer to here. Should this log be done during the compaction in tsdb?